src/core/Akka/Dispatch/Dispatchers.cs

```diff
@@ -137,6 +138,54 @@ public override void Shutdown()
         }
     }
 
+    internal sealed class ChannelExecutor : ExecutorService
+    {
+        private Channel<IRunnable> _channel;
+        private CancellationTokenSource _cts = new CancellationTokenSource();
+
+        private volatile int _readers = 0;
+        private bool HasReaders => _readers > 0;
+
+        public ChannelExecutor(string id, TaskScheduler scheduler, int degreeOfParallelism) : base(id)
+        {
+            _channel = Channel.CreateUnbounded<IRunnable>(new UnboundedChannelOptions()
+            {
+                AllowSynchronousContinuations = true, SingleReader = false, SingleWriter = true
+            });
+
+            for (var i = 0; i < degreeOfParallelism; i++)
+                Task.Run(() => ReadChannel(_channel.Reader));
```

Comment:
I suspect this is the biggest performance problem, how the `ChannelReader<IRunnable>` actually performs consumption via the `ThreadPool` (or in this case, the default `TaskScheduler`) but this is how a lot of other projects and implementations do it. I suspect their performance requirements aren't nearly as high as ours.

Reply:
Take a look at how [`UnboundedChannel` is implemented](https://github.com/dotnet/runtime/blob/79ae74f5ca5c8a6fe3a48935e85bd7374959c570/src/libraries/System.Threading.Channels/src/System/Threading/Channels/UnboundedChannel.cs#L15). The basic approach is that it's a `ConcurrentQueue<>` + a deque with lock for awaiting readers subscription. 

If I understand correctly major incentive over just dispatching tasks to threadpool is control over `degreeOfParallelism`.

You could probably just replace channel with concurrent queue with asynchronous version of AutoResetEvent (it can have very simple construction as it's never awaited more than once).

Reply:
Yeah, the basic goal is to bound the degree of parallelism for multiple workloads in order to prevent starvation of `/system` and `/remote` actors - either the queue has to exist at the level of the OS' thread scheduler (which is what happens today) or it has to exist above it in the form of a virtual work queue. Our working theory is that we're probably better off letting the .NET `ThreadPool` manage the former and for us to apply an abstraction above the latter.